[object] A draft implementation of box style object

jolestar opened the pull request:
## Box Style Move Object 

This PR is part of #21 

In the Sui Move, Object is a Move struct which has some additional requirements:

1. The struct must have the `key` ability.
2. The struct's first field must be `UID` type.

In this PR, I try to introduce a Box-style Move Object. The Object is like a Box and has an ID field. We can put a Move resource to the Box and save it to the Object Store.

```rust
struct Object<T: store> {
        id: UID,
        value: T,
}

public fun new<T: store>(ctx: &mut TxContext, value: T): Object<T>{
    let id = UID {
        id: ID { bytes: tx_context::new_object(ctx) },
    };
    Object<T>{id, value}
}

public fun borrow_value<T: store>(this: &Object<T>): &T{
    &this.value
}

public fun borrow_value_mut<T: store>(this: &mut Object<T>): &mut T{
    &mut this.value
}

public fun transfer<T: store>(obj: Object<T>, recipient: address);
```

The ObjectRef:

```rust
struct ObjectRef< phantom T: store> has store, drop{
    id:  ID,
}

public fun as_object_ref<T: store>(this: &Object<T>): ObjectRef<T>{
    ObjectRef<T>{
        id: this.id.id,
    }
}

public fun borrow<T: store>(this: &ObjectRef<T>): &Object<T>{
        &borrow_object(this.id)
}

native fun borrow_object<T: store>(id: ID): &Object<T>;

```

## Pros

1. Box-style Move Object can be an extension to the Move VM like the Table extension, which does not require modifying the bytecode verifier.

## Cons

1. Developer can not directly define an Object type. (`Article` vs `Object<Article>`).


## TODO

1. Implement the ObjectRef(this will implement in the next PR)

baichuan3 commented on moveos/moveos-stdlib/moveos-stdlib/sources/object.move:
```diff
@@ -141,20 +141,73 @@ module moveos_std::object {
+        &borrow_object(this.id).value
+    }
+
+    public fun borrow_mut_by_ref<T: store>(this: &mut ObjectRef<T>): &mut T{
```
How to get &mut ObjectRef<T>? Is there any security problem to modify T through ObjectRef reference?

jolestar replied:
Yes, I also think about this. Maybe we need to put a field `mutable` to ObjectRef? or define two types of reference: `ObjectRef` and `ObjectMutRef`.

The `ObjectMutRef` can be created by the mut Object.

baichuan3 commented on moveos/moveos-stdlib/moveos-stdlib/sources/object.move:
```diff
@@ -141,20 +141,73 @@ module moveos_std::object {
+    native fun transfer_internal<T: store>(obj: Object<T>, recipient: address);
+
+    // Box style object
+     struct Object<T: store> {
```
In both core Move and Sui Move, the [key ability] denotes a type that can appear as a key in global storage. Does box style Object need the [key ability]  , and the [store ability]?

jolestar replied:
No. `Object` has no `key` or `store`, so the developer can not put Object to global storage(move_to), or embed to another struct. only the `transfer` function can consume the Object.

lshoo commented on moveos/moveos-types/src/object.rs:
```diff
@@ -129,98 +132,120 @@ pub type SequenceNumber = u64;
         }
     }
+
+    pub fn as_move_resource(&self, id: ObjectID) -> Result<Vec<u8>> {
```
I think the move_resource will be a struct(type), so the funtions name`as_move_resource`  will like  `as_resource_bytes`

jolestar replied:
This is a draft method. I try to find a way to map the Move Object struct to Rust Object. Because the Move Object has a generic type parameter `T`, it is hard to implement in Rust.

pause125 replied:
Why need this map( Move object to Rust object)?

jolestar replied:
We need parse the Move object in Rust code.

jolestar replied:
I rename this function to `as_object_argument`

jolestar replied:
> I think there's no need to parse the whole Object. Maybe parse the Object::id field is enough?

Yes.

Rust Move Object
```
struct MoveObject{
  id: ObjectID,
  value: Vec<u8>,
}
```

Move Object
```
struct Object<T>{
   id: address,
   value: T,
}
```

I can not find a simple method to make rust `bcs::to_bytes(&move_object)` == move `bcs::to_bytes(&object)`

jolestar commented:
I remove the ObjectRef, please help review @baichuan3 @pause125

baichuan3 commented:
> I remove the ObjectRef, please help review @baichuan3 @pause125

Is object ref implemented later, or does this feature no longer exist?

jolestar commented:
> > I remove the ObjectRef, please help review @baichuan3 @pause125
> 
> Is object ref implemented later, or does this feature no longer exist?

The ObjectRef design #51

baichuan3 commented on moveos/moveos-stdlib/src/natives/moveos_stdlib/object.rs:
```diff
@@ -153,6 +153,23 @@ pub fn native_share_object(
 
+// ObjectBox keep same layout with Object struct in Move
+pub(crate) struct ObjectBox {
+    pub id: AccountAddress,
```
If the account address is used directly for the object id, it means that if the recipient address passed in when transferring is the object id, because it cannot be verified, the owner will become the object id and the object will never be accessible again.

jolestar replied:
It is like transferring Token to a no exists account.

baichuan3 commented on moveos/moveos-stdlib/src/natives/moveos_stdlib/any_table/mod.rs:
```diff
@@ -0,0 +1,665 @@
+// SPDX-License-Identifier: Apache-2.0
+
+/// A native Table implementation for save any type of value.
+/// Refactor from https://github.com/rooch-network/move/blob/5b413d009515ad1144042ded27cbe9bd702aaad7/language/extensions/move-table-extension/src/lib.rs#L4
```
any table is used to implement storage abstraction?

jolestar replied:
yes

baichuan3 commented:
wait for further implementation of object and then continuing the review

baichuan3 commented on moveos/moveos-stdlib/moveos-stdlib/sources/object.move:
```diff
@@ -2,159 +2,88 @@
+    #[private_generic(T)]
+    /// Create a new object, the object is owned by `owner`
+    /// The private generic is indicate the T should be defined in the same module as the caller. This is ensured by the verifier.
+    public fun new<T: key>(ctx: &mut TxContext, owner: address, value: T): Object<T>{
```
When the address corresponding to the signer in the owner and tx_context is inconsistent, it is equivalent to implicitly operating the transfer after creating the object, and the owner passes the parameter to become the owner of the object?

jolestar replied:
The object owner can be a resource account, so let the caller decide who is owner.

baichuan3 commented on moveos/moveos-stdlib/moveos-stdlib/sources/object.move:
```diff
@@ -2,159 +2,88 @@
+    
+    #[private_generic(T)]
+    /// Add object to object store
+    public fun add<T: key>(this: &mut ObjectStore, obj: Object<T>) {
```
why T need ability `key`? Does object stroage is independent of global storage ?

jolestar replied:
I want to reuse the `key` ability to distinguish the object type and other types.

In Sui object, the object must have the `key` ability, and if have both `key` and `store`, the object can be public transfer.

jolestar commented:
1. implements raw_table (formerly any_table)
2. implements Table, TypeTable, ObjectStorage based on raw_table
3. Implemented AccountStorage based on TypeTable, Table and ObjectStorage, and implemented alternative methods for instructions such as move_to|move_from|borrow_global in AccountStorage.
4. Unify the global storage command and table with raw_table.

baichuan3 commented:
LGTM

jolestar commented:
TODO:

1. Do some cleanup
2. Disable move_to|move_from|borrow_global or use same GlobalValue with table.
